k-distribution/k-tutorial/2_intermediate/01_macros/README.md

```diff
@@ -4,6 +4,151 @@ copyright: Copyright (c) Runtime Verification, Inc. All Rights Reserved.
+another configuration;
+2. Function rules, which define the behavior of a function written over
+arbitrary input and output types; and
+3. Simplification rules, which describe ways in which the symbolic execution
+engine ought to simplify terms containing symbolic values.
+
+This lesson introduces three more types of rules, of which the first are
+**macros**. A production is a macro if it has the `macro` attribute, and all
+rules whose top symbol on the left hand side is a macro are **macro rules**
+which define the behavior of the macro. Like function rules and simplification
+rules, macro rules do not participate in cell completion. However, unlike
+function rules and simplification rules, macro rules are applied **statically**
+before rewriting begins, and the macro symbol is expected to no longer appear
+in the initial configuration for rewriting once all macros in that
+configuration are rewritten.
+
+The rationale behind macros is they allow you to define one piece of syntax
+in terms of another piece of syntax without any runtime overhead associated
+with the cost of rewriting one to the other. For example, in a language with
+`if` statements and curly braces, I could write the following fragment
+(`lesson-01.k`):
+
+```k
+module LESSON-01
+  syntax Stmt ::= "if" "(" Exp ")" Stmt             [macro]
+                | "if" "(" Exp ")" Stmt "else" Stmt
+                | "{" Stmts "}"
+  syntax Stmts ::= List{Stmt,""}
+
+  rule if ( E ) S => if ( E ) S else { }
+endmodule
+```
+
+In this example, we see that an if statement without an `else` clause is
+defined in terms of one with an `else` clause. As a result, we would only
+need to give a single rule for how to rewrite `if` statements, rather than
+two separate rules for two types of `if` statements. This is a common pattern
+for dealing with program syntax that contains an optional component to it.
+
+It's worth noting, by default, macros are not applied recursively. To be more
+precise, by default a macro that arises as a result of the expansion of the
+same macro is not rewritten further. This is primarily to simplify the macro
+expansion process and reduce the risk that improperly defined macros will
+lead to non-terminating behavior.
+
+It is possible, however, to tell K to expand a macro recursively. To do this,
+simply replace the `macro` attribute with the `macro-rec` attribute. Note that
+K does not do any kind of checking to ensure termination here, so it is
+important that rules be defined correctly to always terminate, otherwise the
+macro expansion phase will run forever. In practice, the types of macros
+usually written are very easy to ensure this property on, however.
+
+### Exercise
+
+Using a `Nat` sort containing the constructors `0` and `S` (i.e., a
+peano-style axiomatization of the natural numbers where `S(N) = N + 1`,
+`S(S(N)) = N + 2`, etc), write a macro that will compute the sum of two
+numbers.
```

Comment:
I think it's worth linking to an article on Peano arithmetic, maybe here: https://wiki.haskell.org/Peano_numbers